specs/006-mcp-server-architecture/research.md

```diff
@@ -0,0 +1,576 @@
+        span.add_event("format_detected", {"format": "Intan"})
+```
+
+**Context Propagation**:
+- Automatic propagation in async contexts
+- Manual propagation for agent invocations
+- Trace ID injection into logs for correlation
+
+**Export Strategy**:
+- OTLP exporter to collector (production)
+- Console exporter for development
+- Batch span processor for performance
+
+### Alternatives Considered
+
+1. **Jaeger client directly**
+   - Rejected: Vendor lock-in, no metrics/logs, manual instrumentation
+
+2. **AWS X-Ray**
+   - Rejected: Cloud vendor lock-in, poor local development experience
+
+3. **Datadog APM directly**
+   - Rejected: Expensive, vendor lock-in, OpenTelemetry can export to Datadog anyway
+
+4. **Custom logging-based tracing**
+   - Rejected: No distributed context, manual correlation, poor tooling
+
+---
+
+## 5. Format Detection and File Analysis
+
+### Decision Made
+**Multi-layered detection** using file signatures, magic bytes, and NeuroConv interface mapping
+
+### Rationale
+- Neuroscience formats (25+) have varying levels of structure
+- No single detection method works for all formats
+- Layered approach increases confidence and handles edge cases
+- Leverages NeuroConv's existing interface definitions
+
+### Detection Layers
+
+**Layer 1: File Extension Analysis**
+- Fast initial filter using pathlib
+- Confidence scoring based on extension uniqueness
+- Handles common formats (.nex, .nwb, .smrx)
+
+**Layer 2: Magic Bytes Detection**
+- Read first 512 bytes of binary files
+- Compare against format signature database
+- Handles HDF5-based formats (NWB, Allen Brain Observatory)
+
+**Layer 3: Directory Structure Pattern Matching**
+- Analyze file hierarchy for known patterns
+- SpikeGLX (*.ap.bin + *.ap.meta), Intan (multiple .rhd files)
+- Regex patterns for common structures
+
+**Layer 4: Header Parsing**
+- Parse format-specific headers for metadata
+- Validate expected fields and structure
+- Extract version information
+
+**Layer 5: NeuroConv Interface Mapping**
+- Map detected signatures to NeuroConv data interfaces
+- Use confidence weights from previous layers
+- Select optimal interface for conversion
+
+### Format Signature Database
+- JSON configuration file with format definitions
+- Fields: name, extensions, magic_bytes, directory_patterns, neuroconv_interface
+- Extensible for new formats via configuration
+
+### Alternatives Considered
+
+1. **python-magic library only**
+   - Rejected: Generic format detection, lacks neuroscience-specific knowledge
+
+2. **File extension only**
+   - Rejected: Unreliable, files may lack extensions or have wrong ones
+
+3. **Deep file content analysis**
+   - Rejected: Too slow for initial detection, better as validation step
+
+4. **Machine learning classifier**
+   - Rejected: Overkill, requires training data, less explainable
+
+---
+
+## 6. Agent Communication Patterns
+
+### Decision Made
+**Async RPC pattern** with request/response queuing
```

Comment:
this seems a little excessive but I guess it will be helpful to build experience with the technology.